tardis/utilities/utils.py

```diff
@@ -40,21 +39,34 @@ def htcondor_cmd_option_formatter(options: AttributeDict) -> str:
     return cmd_option_formatter(options, prefix="-", separator=" ")
 
 
-def htcondor_csv_parser(
-    htcondor_input: str,
-    fieldnames: [List, Tuple],
-    delimiter: str = "\t",
-    replacements: dict = None,
+def csv_parser(
+    input_csv, fieldnames, delimiter="\t", replacements=None, skipinitialspace=False
 ):
+    """
+    Parses CSV formatted input
+
+    :param input_csv: CSV formatted input
+    :type input_csv: str
+    :param fieldnames: corresponding field names
+    :type fieldnames: str
+    :param delimiter: delimiter between entries
+    :type delimiter: char
+    :param replacements: fields to be replaced
+    :type replacements: dict
+    """
     replacements = replacements or {}
-    with StringIO(htcondor_input) as csv_input:
-        cvs_reader = csv.DictReader(
-            csv_input, fieldnames=fieldnames, delimiter=delimiter
+    with StringIO(input_csv) as csv_input:
+        csv_reader = csv.DictReader(
+            csv_input,
+            fieldnames=fieldnames,
+            delimiter=delimiter,
+            skipinitialspace=skipinitialspace,
         )
-        for row in cvs_reader:
+        for row in csv_reader:
             yield {
                 key: value if value not in replacements.keys() else replacements[value]
                 for key, value in row.items()
+                if key is not None
```

Comment:
It seems that if the delimiters are whitespaces and if in addition a row ends with whitespaces, it adds another column at the end with no information, which is then represented as `(None, [''])`. Therefore I've added this "safety net". I don't know how to fix this a level up (in the adapter) without jumping through hoops. Do you have any opinions on this?

Reply:
Would another delimiter work? You can chose it by using `--format` instead of `--Format`.

```
sinfo --format="..|..|..|" ...
```

Reply:
This is unfortunately not an option. It is necessary to use `--Format` because the field `AllocMem` is not available when using `--format`.

Reply:
Looking at the sinfo manual it seems that at least having a comma separated list should be possible (from the manual for the -O/--Format option for sinfo "Requests a comma separated list of job information to be displayed."), would this be an option?
However, I might be misunderstanding the manual on this point and don't have access to a system with slurm to verify this is possible.

Reply:
This sentence in the documentation is very misleading. I think it means that the way you request this information is via a comma separated list:
```
sinfo  --Format="statelong,cpusstate,allocmem,memory,features,nodehost,reason"
```
The output is then
```
idle                0/40/0/40           0                   100000              nemo-XXXXX        host-1     none
...
```
I could maybe try to first chomp off trailing whitespace from each row before I feed it to the `csv_parser`, but is it really worth the effort?

Reply:
Honestly, I would prefer to solve the problem before parsing the csv. I would suggest to use the following code to do that.

```python
def csv_parser(
    input_csv, fieldnames, delimiter="\t", replacements=None, skipinitialspace=False, skiptrailingspace=False
): 
...
if skiptrailingspace:
    input_csv = "\n".join((line.strip() for line in input_csv.splitlines()))
...
```

Reply:
Thanks!

Reply:
Interestingly, the py3.9 builds are failing because of this change:
```
    input_csv = "\n".join((line.strip() for line in input_csv.splitlines()))

AttributeError: '_asyncio.Future' object has no attribute 'splitlines'
```
Do you have any idea what might be the reason? My knowledge of asyncio is rather limited.

Reply:
Yes. This is a problem how we are mocking the async call in the unittests. Newer Python versions autodetect async coroutines and use `AsyncMock` (https://docs.python.org/3/library/unittest.mock.html#unittest.mock.AsyncMock) instead of a `MagicMock`. Our code assumes that `MagicMock` is used and mocks the async call itself. So, we end up with two nested async calls in newer Python versions, leading to the error above. The solution is to force the test to use a `MagicMock`. Could you try to change the code, like in https://github.com/MatterMiners/tardis/commit/cd639e8cdf9fe56d5f53526f459392904a414437, please?

Reply:
Thanks a lot, this seems to do the trick!